### PR TITLE
New resource: `azurerm_container_connected_registry`

### PR DESCRIPTION
**Pull request by magodo**
This PR implements a new resource: `azurerm_container_connected_registry`.

## Reference

https://docs.microsoft.com/en-us/azure/container-registry/intro-connected-registry

### CODE REVIEW

File: internal/services/containers/container_connected_registry_resource.go [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+					string(containerregistry.ConnectedRegistryModeMirror),
+					string(containerregistry.ConnectedRegistryModeReadOnly),
+					string(containerregistry.ConnectedRegistryModeReadWrite),
+					string(containerregistry.ConnectedRegistryModeRegistry),
+				},
+				false,
+			),
+		},
+
+		"notification_list": {
```
Comment (katbyte) [inline internal/services/containers/container_connected_registry_resource.go:104]: list is redundant, also what type of destination is expected here?
```suggestion
		"notification": {
```

also crazy idea, would it make sense to break this out into blocks validating each component? open to input on that

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @magodo, in addition to my comments inline looks like the tests are failing:
```
------- Stdout: -------
=== RUN   TestAccContainerConnectedRegistry_mirror
=== PAUSE TestAccContainerConnectedRegistry_mirror
=== CONT  TestAccContainerConnectedRegistry_mirror
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: checking for presence of existing Container Connected Registry: (Connected Registry Name "testacccrc220322210138990297" / Registry Name "testacccr220322210138990297" / Resource Group "acctestRG-acr-220322210138990297"): containerregistry.ConnectedRegistriesClient#Get: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: error response cannot be parsed: "\"The resource type 'connectedRegistries' is not supported in this region.\"" error: json: cannot unmarshal string into Go value of type azure.RequestError
        
          with azurerm_container_connected_registry.test,
          on terraform_plugin_test.tf line 40, in resource "azurerm_container_connected_registry" "test":
          40: resource "azurerm_container_connected_registry" "test" {
        
        checking for presence of existing Container Connected Registry: (Connected
        Registry Name "testacccrc220322210138990297" / Registry Name
        "testacccr220322210138990297" / Resource Group
        "acctestRG-acr-220322210138990297"):
        containerregistry.ConnectedRegistriesClient#Get: Failure responding to
        request: StatusCode=400 -- Original Error: autorest/azure: error response
        cannot be parsed: "\"The resource type 'connectedRegistries' is not supported
        in this region.\"" error: json: cannot unmarshal string into Go value of type
        azure.RequestError
--- FAIL: TestAccContainerConnectedRegistry_mirror (172.11s)
FAIL
```

File: internal/services/containers/container_connected_registry_resource.go (hunk 1 above)
Comment (magodo) [reply internal/services/containers/container_connected_registry_resource.go]: @katbyte Good idea, I'll make the change.

**Comment by magodo** [comment]
@katbyte I've now refactored the `notification` from list of string into blocks. Please take another look! About the test failures, from the error message, it seems you are running it in a region that doesn't support the connected registry? At least west europe should support that.

## Test

```shell
terraform-provider-azurerm on  acr_conn_reg via 🐹 v1.18 took 6s
💤  TF_ACC=1 go test -v -timeout=20h ./internal/services/containers  -run='TestAccContainerConnectedRegistry_'
=== RUN   TestAccContainerConnectedRegistry_basic
=== PAUSE TestAccContainerConnectedRegistry_basic
=== RUN   TestAccContainerConnectedRegistry_complete
=== PAUSE TestAccContainerConnectedRegistry_complete
=== RUN   TestAccContainerConnectedRegistry_update
=== PAUSE TestAccContainerConnectedRegistry_update
=== RUN   TestAccContainerConnectedRegistry_mirror
=== PAUSE TestAccContainerConnectedRegistry_mirror
=== RUN   TestAccContainerConnectedRegistry_registry
=== PAUSE TestAccContainerConnectedRegistry_registry
=== RUN   TestAccContainerConnectedRegistry_cascaded
=== PAUSE TestAccContainerConnectedRegistry_cascaded
=== RUN   TestAccContainerConnectedRegistry_requiresImport
=== PAUSE TestAccContainerConnectedRegistry_requiresImport
=== CONT  TestAccContainerConnectedRegistry_basic
=== CONT  TestAccContainerConnectedRegistry_registry
=== CONT  TestAccContainerConnectedRegistry_update
=== CONT  TestAccContainerConnectedRegistry_mirror
=== CONT  TestAccContainerConnectedRegistry_complete
=== CONT  TestAccContainerConnectedRegistry_requiresImport
=== CONT  TestAccContainerConnectedRegistry_cascaded
--- PASS: TestAccContainerConnectedRegistry_complete (158.29s)
--- PASS: TestAccContainerConnectedRegistry_mirror (164.51s)
--- PASS: TestAccContainerConnectedRegistry_requiresImport (189.01s)
--- PASS: TestAccContainerConnectedRegistry_cascaded (222.85s)
--- PASS: TestAccContainerConnectedRegistry_registry (223.45s)
--- PASS: TestAccContainerConnectedRegistry_basic (774.41s)
--- PASS: TestAccContainerConnectedRegistry_update (1015.60s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/containers    1015.618s
```

**Comment by katbyte** [comment]
@magodo its just a standards TC run:

```
------- Stdout: -------
=== RUN   TestAccContainerConnectedRegistry_update
=== PAUSE TestAccContainerConnectedRegistry_update
=== CONT  TestAccContainerConnectedRegistry_update
    testcase.go:110: Step 1/6 error: Error running apply: exit status 1
        
        Error: checking for presence of existing Container Connected Registry: (Connected Registry Name "testacccrc220407005833738639" / Registry Name "testacccr220407005833738639" / Resource Group "acctestRG-acr-220407005833738639"): containerregistry.ConnectedRegistriesClient#Get: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: error response cannot be parsed: {"\"The resource type 'connectedRegistries' is not supported in this region.\"" '\x00' '\x00'} error: json: cannot unmarshal string into Go value of type azure.RequestError
        
          with azurerm_container_connected_registry.test,
          on terraform_plugin_test.tf line 45, in resource "azurerm_container_connected_registry" "test":
          45: resource "azurerm_container_connected_registry" "test" {
        
        checking for presence of existing Container Connected Registry: (Connected
        Registry Name "testacccrc220407005833738639" / Registry Name
        "testacccr220407005833738639" / Resource Group
        "acctestRG-acr-220407005833738639"):
        containerregistry.ConnectedRegistriesClient#Get: Failure responding to
        request: StatusCode=400 -- Original Error: autorest/azure: error response
        cannot be parsed: {"\"The resource type 'connectedRegistries' is not
        supported in this region.\"" '\x00' '\x00'} error: json: cannot unmarshal
        string into Go value of type azure.RequestError
--- FAIL: TestAccContainerConnectedRegistry_update (119.03s)
FAIL
```

primary location is west eu

**Comment by magodo** [comment]
@katbyte I assume that is because the [dynamic test](https://github.com/hashicorp/terraform-provider-azurerm/blob/87a2de01819597f8dab6bd54797d1f2e7a85fdef/internal/acceptance/data.go#L81) is enabled for this run:

image

File: website/docs/r/container_connected_registry.html.markdown [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+* `client_token_ids` - (Optional) Specifies a list of IDs of Container Registry Tokens, which are meant to be used by the clients to connect to the Connected Registry.
+
+* `log_level` - (Optional) The verbosity of the logs. Possible values are `None`, `Debug`, `Information`, `Warning` and `Error`.
+
+* `mode` - (Optional) The mode of the Connected Registry. Possible values are `Mirror`, `ReadOnly`, `ReadWrite` and `Registry`. Changing this forces a new Container Connected Registry to be created.
+
+* `notification` - (Optional) One or more `notification` blocks as defined below.
+i
+Specifies a list of notification subscriptions of the Connected Registry. Each notification is in the following format: `<artifacte>[:<tag>|@<digest>]:<action>`, where action can be one of `push`, `delete` and `*` (i.e. any).
```
Comment (katbyte) [inline website/docs/r/container_connected_registry.html.markdown:83]: typoy
```suggestion
* `notification` - (Optional) One or more `notification` blocks as defined below.

```

**Review by katbyte — APPROVED** [review]
Thanks @magodo - LGTM aside from 1 comment

**Pull request merged**